src/Routing/Router.php

```diff
@@ -650,18 +650,29 @@ public static function reverseToArray($params): array
         }
         $pass = $params['pass'] ?? [];
 
+        $template = $params['_matchedRoute'] ?? null;
         unset(
             $params['pass'],
             $params['_matchedRoute'],
             $params['_name']
         );
-        foreach ($pass as $i => $passedValue) {
-            // Remove passed values that are also route keys
-            if (in_array($passedValue, $params, true)) {
-                unset($pass[$i]);
+        $route = null;
+        if ($template) {
+            // Locate the route that was used to match this route
+            // so we can access the pass parameter configuration.
+            foreach (static::getRouteCollection()->routes() as $maybe) {
+                if ($maybe->template === $template) {
+                    $route = $maybe;
+                    break;
+                }
             }
         }
-        $params = array_merge($params, array_values($pass));
+        if ($route) {
```

Comment:
Should there be an exception instead when no route matching the template is found? Trying to reverse an array which doesn't match to any existing routes doesn't seem very useful.

Reply:
I was concerned about breaking an application that removed the `_matchingRoute` key

Reply:
True, the returned array would be "unusable", so I guess this is fine.